sympy/matrices/common.py

```diff
@@ -2606,6 +2606,9 @@ def __repr__(self):
     def shape(self):
         return (self.rows, self.cols)
 
+    def as_immutable(self): # this is needed here ONLY FOR TESTS.
```

Comment:
Which tests is this needed for? Would it be better to use an actual `Matrix` in those tests?

Reply:
This snuck in unintentionally and actually it is a preparation for the next set of modifications which requires some explaining, but as for why this is in `_MinimalMatrix` it is for the tests in `test_commonmatrix.py` which create classes like `class ShapingOnlyMatrix(_MinimalMatrix, MatrixShaping)`. I also find these tests strange and wonder why they are not just using a normal Matrix as well but I did not want to open up that can of worms at this point.

Now on to explaining the `as_immutable` there, I will give you the full background since you seem to know the matrix module and can probably give good criticism. We had a long back-and-forth with @oscarbenjamin and @sylee957 on matrix caching and how it can not be done currently since both mutable and immutable matrices share the same function entry points. @oscarbenjamin eventually convinced me that the best method to go about implementing this is to have important internal matrix functions like `rref` or `eigenvals` operate on immutable matrices and cache those via `cacheit`. casting mutable to immutable matrices for the operation and back to mutable after from the matrix methods at least.

This is actually one of the other reasons to pull out the actual implementations of the functions so that the method wrappers and eventual top-level wrappers can coerce appropriately and the internal functions can eventually expect only immutable types and can be cached correctly with `cacheit` (which would mean that even mutable matrices would benefit). This is not the end of the story though, in solvers there are two classes `RawMatrix` and `NewMatrix` which work with elements subclassed from `CantSympify` but I think I have this solved and caching can actually now be added to matrices.